common/src/main/kotlin/com/gitlab/kordlib/common/entity/Presence.kt

```diff
@@ -1,62 +1,92 @@
-        val email: JsonElement? = null
-)
+        val id: Snowflake,
+        val details: JsonObject,
+) {
+
+    internal object Serializer : KSerializer<DiscordPresenceUser> {
+
+        override val descriptor: SerialDescriptor = buildClassSerialDescriptor("Kord.DiscordPresenceUser") {
+            element<Snowflake>("id")
+            element<JsonElement>("details")
+        }
+
+        override fun deserialize(decoder: Decoder): DiscordPresenceUser {
+            val jsonDecoder = decoder as? JsonDecoder ?: error("Can be deserialized only by JSON")
+            val json = jsonDecoder.decodeJsonElement().jsonObject
+            val id = Snowflake(json.getValue("id").jsonPrimitive.content)
+            val details = json.toMutableMap()
+            details.remove("id")
+
+            return DiscordPresenceUser(id, JsonObject(details))
+        }
+
+        override fun serialize(encoder: Encoder, value: DiscordPresenceUser) {
+            val jsonEncoder = encoder as? JsonEncoder ?: error("Can be serialized only by JSON")
+            val details = value.details.toMutableMap()
+            details["id"] = JsonPrimitive(value.id.asString)
+
+            jsonEncoder.encodeJsonElement(JsonObject(details))
+        }
+    }
+
+}
 
 @Serializable
-data class DiscordClientStatus(val desktop: Status? = null, val mobile: Status? = null, val web: Status? = null)
+data class DiscordClientStatus(
+        val desktop: Optional<PresenceStatus> = Optional.Missing(),
+        val mobile: Optional<PresenceStatus> = Optional.Missing(),
+        val web: Optional<PresenceStatus> = Optional.Missing(),
+)
+
+@Serializable(with = PresenceStatus.StatusSerializer::class)
+sealed class PresenceStatus(val value: String) {
 
-@Serializable(with = Status.StatusSerializer::class)
-enum class Status {
-        /** The default code for unknown values. */
-        Unknown,
-        Online, DnD, Idle, Invisible, Offline;
+    class Unknown(value: String) : PresenceStatus(value)
+    object Online : PresenceStatus("online")
+    object Idle : PresenceStatus("idle")
+    object DoNotDisturb : PresenceStatus("dnd")
 
-        companion object StatusSerializer : KSerializer<Status> {
-                override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Status", PrimitiveKind.STRING)
+    companion object StatusSerializer : KSerializer<PresenceStatus> {
+        override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Kord.ClientStatus", PrimitiveKind.STRING)
 
-                override fun deserialize(decoder: Decoder): Status {
-                        val name = decoder.decodeString()
-                        return values().firstOrNull { it.name.toLowerCase() == name } ?: Unknown
-                }
+        override fun deserialize(decoder: Decoder): PresenceStatus = when (val value = decoder.decodeString()) {
+            "online" -> Online
+            "idle" -> Idle
+            "dnd" -> DoNotDisturb
+            else -> Unknown(value)
```

Comment:
invisble and offline are missing

Reply:
Technically so. Yet the gateway will not emit presences for those who appear 'offline' and does not reveal invisible statuses either. I'll add them for completeness though.